### PR TITLE
EmacsRequirement

### PR DESCRIPTION
**Pull request by dunn**
@xu-cheng floated the idea in #39266, so I figured I'd put this forward as a possibility.

I'm still getting familiar with the codebase and Ruby in general, so lmk if there are any obvious mistakes.

### CODE REVIEW

**Comment by dunn** [comment]
Pulling in some comments from #39266: specific Emacs requirements might be too much of a corner case.  The formulae I found that can currently benefit are just four:
- [`magit.rb`](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/magit.rb) 
- [`cask.rb`](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/cask.rb) 
- [`mu.rb`](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/mu.rb)
- [`proof-general.rb`](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/proof-general.rb#L28)

But in the future this might allow fewer formula to require brewed Emacs.

File: Library/Homebrew/requirements/emacs_requirement.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+  default_formula "emacs"
+
+  def initialize(tags)
+    @version = tags.shift if /\d+\.*\d*/ === tags.first
+    raise "Specify a version for EmacsRequirement" unless @version
+    super
+  end
+
+  satisfy do
```
Comment (xu-cheng) [inline Library/Homebrew/requirements/emacs_requirement.rb]: `satisfy :build_env => false do`

File: Library/Homebrew/requirements/emacs_requirement.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    emacs_version = `emacs --batch --eval "(princ emacs-version)"`.to_f
```
Comment (xu-cheng) [inline Library/Homebrew/requirements/emacs_requirement.rb]: Need to check whether emacs is installed first.

File: Library/Homebrew/requirements/emacs_requirement.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    emacs_version >= @version.to_f
```
Comment (xu-cheng) [inline Library/Homebrew/requirements/emacs_requirement.rb]: Using `Version.new(version.to_s)`

**Comment by xu-cheng** [comment]
There also should be a `env` block to change PATH.

I suggest you to have look at [RubyRequirement](https://github.com/Homebrew/homebrew/blob/master/Library/Homebrew/requirements/ruby_requirement.rb) as example.

**Comment by dunn** [comment]
Is the `env` block to ensure that the same `emacs` that's used to test the requirement is then used in the build process?

**Comment by xu-cheng** [comment]
Actually, the way requirement works is to only check first program found in the PATH. So yes, they are the same emacs. Also could you add a `inspect` method as well? Thanks.

**Comment by dunn** [comment]
For `inspect`, X11 uses `min_version=` while Java and Ruby just go with `version=`. Any preference here?

**Comment by xu-cheng** [comment]
I think both are OK. But let's go with `version=`

**Comment by xu-cheng** [comment]
LGTM. Will let @Homebrew/owners to have a look.

**Comment by xu-cheng** [comment]
@dunn :ship:ed. Can you apply this change to the formulae you listed above? Thanks.

**Pull request closed without merging**